File: packages/object-detection/ios/Plugin/Classes/Results/ProcessImageResult.swift
import Foundation
import Capacitor
import MLKitObjectDetection

@objc public class ProcessImageResult: NSObject, Result {
    let objects: [Object]

    init(objects: [Object]) {
        self.objects = objects
    }

    @objc public func toJSObject() -> AnyObject {
        let detectedObjectsResult = createDetectedObjectsResult()

        var result = JSObject()
        result["detectedObjects"] = detectedObjectsResult

        return result as AnyObject
    }

    private func createDetectedObjectsResult() -> JSArray {
        var result = JSArray()

        for object in objects {
            let detectedObjectResult = createDetectedObjectResult(object)
            result.append(detectedObjectResult)
        }

        return result
    }

    private func createDetectedObjectResult(_ object: Object) -> JSObject {
        var result = JSObject()
        result["boundingBox"] = createBoundingBoxResult(object.frame)
        if let trackingID = object.trackingID {
            result["trackingId"] = trackingID.intValue
        }
        result["labels"] = createLabelsResult(object.labels)
        return result
    }

    private func createBoundingBoxResult(_ frame: CGRect) -> JSObject {
        var result = JSObject()
        result["left"] = Int(frame.minX)
        result["top"] = Int(frame.minY)
        result["right"] = Int(frame.maxX)
        result["bottom"] = Int(frame.maxY)
        return result
    }

    private func createLabelsResult(_ labels: [ObjectLabel]) -> JSArray {
        var result = JSArray()

        for label in labels {
            var labelResult = JSObject()
            labelResult["index"] = label.index
            labelResult["text"] = label.text
            labelResult["confidence"] = label.confidence
            result.append(labelResult)
        }

        return result
    }
}

File: packages/object-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/objectdetection/classes/results/ProcessImageResult.java
package io.capawesome.capacitorjs.plugins.mlkit.objectdetection.classes.results;

import android.graphics.Rect;
import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.vision.objects.DetectedObject;
import io.capawesome.capacitorjs.plugins.mlkit.objectdetection.interfaces.Result;
import java.util.List;

public class ProcessImageResult implements Result {

    @NonNull
    private final List<DetectedObject> detectedObjects;

    public ProcessImageResult(@NonNull List<DetectedObject> detectedObjects) {
        this.detectedObjects = detectedObjects;
    }

    @Override
    @NonNull
    public JSObject toJSObject() {
        JSArray detectedObjectsResult = this.createDetectedObjectsResult();

        JSObject result = new JSObject();
        result.put("detectedObjects", detectedObjectsResult);
        return result;
    }

    private JSArray createDetectedObjectsResult() {
        JSArray result = new JSArray();
        for (DetectedObject detectedObject : detectedObjects) {
            JSObject detectedObjectResult = this.createDetectedObjectResult(detectedObject);
            result.put(detectedObjectResult);
        }
        return result;
    }

    private JSObject createDetectedObjectResult(@NonNull DetectedObject detectedObject) {
        JSObject result = new JSObject();
        result.put("boundingBox", this.createBoundingBoxResult(detectedObject.getBoundingBox()));
        Integer trackingId = detectedObject.getTrackingId();
        if (trackingId != null) {
            result.put("trackingId", trackingId);
        }
        result.put("labels", this.createLabelsResult(detectedObject.getLabels()));
        return result;
    }

    private JSObject createBoundingBoxResult(@NonNull Rect boundingBox) {
        JSObject result = new JSObject();
        result.put("left", boundingBox.left);
        result.put("top", boundingBox.top);
        result.put("right", boundingBox.right);
        result.put("bottom", boundingBox.bottom);
        return result;
    }

    private JSArray createLabelsResult(@NonNull List<DetectedObject.Label> labels) {
        JSArray result = new JSArray();
        for (DetectedObject.Label label : labels) {
            JSObject labelResult = new JSObject();
            labelResult.put("index", label.getIndex());
            labelResult.put("text", label.getText());
            labelResult.put("confidence", label.getConfidence());
            result.put(labelResult);
        }
        return result;
    }
}
